### PR TITLE
Routing logic refactor; Stateless CRUD APIs

### LINKED ISSUE #861: Force Sync API/Command
In the unusual case where configs become conflicted, we should provide a force sync to re-index internal configurations, endpoints, conditions and time series storage

### LINKED ISSUE #943: Config & Routing refactor
- Wrap components of opni's AlertManager config into interfaces
- Allow for broadcasting channels for pushing internal messages
- Allow for syncing user configs 
- Better embedded server inside the AlertManager cluster

### LINKED ISSUE #944: Alerting : sync production configs

### PR DESCRIPTION
**Pull request by alexandreLamarre**
Config Routing refactor ( Closes #943 ) :
- [x] Opni managed sub routes : namespaces, including a catch all finalizer hook for all namespaces
- [X] Opni default namespace for alarms/messages
- [x] Internal routing is merged into a single object which is responsible for all updates to the config
- [x] Storage Node CRUDs the internal config(s)
- [x] Opni embedded server uses GRPC
- [x] Opni embedded server uses PutConfig(config) method instead of areConfigsEquals
- [x] Internal routing and AM config no longer go through k8s crds / cluster drivers

<hr/>
Allow for self healing / user prompted alerting sync to external depdencies, e.g. storage, watchers, cortex : Closes #861

- [x] Implement Force Sync on Storage Node
- [x] Implement Force Sync on Embedded Server

<hr/>

Allow for feature parity so that we can sync the user's production configs : Closes #944 
- [x] Production/user config syncer
- [x] Have the ruler label all alerts that should be sent to the user/production config

Upgrade to AlertManager v0.25:  https://github.com/rancher/opni/issues/945

### CODE REVIEW

**Comment by alexandreLamarre** [comment]
Need to have some test coverage before marking ready for review for the new features of the refactor: 
- routing integrations for router -- making sure they uniquely dispatch to the correct endpoints 
- syncer server + gateway reconciling user configs -- making sure there are no sync issues or storage issues on the alerting gateway with integration tests

**Comment by alexandreLamarre** [comment]
Fixed some data-races specific to the test scenarios code, tweaked some default/finalizer hook throttling settings for better routing, fixed a bug where finalizer hooks could catch messages more than once.

Added some in depth tests to make sure the active state is correctly set in routing and that that activate state is fully routed to the endpoints from the opinionated config construction by `OpniRouter` implementations

**Comment by alexandreLamarre** [comment]
Realized that there needed to be a router-specific abstraction where we can map a `conditionId` (opaque identifier persisted on the conditions server to uniquely identify  a configuration) to the underlying alertmanager prometheus labels that route it to its destination in order to query its status in the alerting backend.

As a result added `	ConditionRoutingLabels(conditionId string) []*labels.Matcher` to the OpniRouting interface.

File: pkg/apis/alerting/v1/validate.go [hunk 1]
```diff
@@ -18,6 +19,22 @@ func validComparionOperator(op string) error {
 	return nil
 }
 
+func (s *SyncerConfig) Validate() error {
+	if s.GatewayJoinAddress == "" {
+		return validation.Error("GatewayJoinAddress must be set")
```
Comment (kralicky) [inline pkg/apis/alerting/v1/validate.go:24]: missing field errors should all be of the form
`fmt.Errorf("%w: %s", validation.ErrMissingRequiredField, <field name>)`

File: plugins/alerting/pkg/alerting/api.go [hunk 2]
```diff
@@ -3,17 +3,22 @@ package alerting
...
 )
 
 // --- Trigger ---
-
 func (p *Plugin) TriggerAlerts(ctx context.Context, req *alertingv1.TriggerAlertsRequest) (*alertingv1.TriggerAlertsResponse, error) {
 	lg := p.Logger.With("Handler", "TriggerAlerts")
 	lg.Debugf("Received request to trigger alerts  on condition %s", req.GetConditionId())
 	lg.Debugf("Received alert annotations : %s", req.Annotations)
-
+	if err := req.Validate(); err != nil {
```
Comment (kralicky) [inline plugins/alerting/pkg/alerting/api.go:19]: Should move this check above the log statements

File: plugins/alerting/pkg/alerting/api.go [hunk 3]
```diff
@@ -24,13 +29,26 @@ func (p *Plugin) TriggerAlerts(ctx context.Context, req *alertingv1.TriggerAlert
...
+	}
+
+	if _, ok := req.Labels[shared.BackendConditionIdLabel]; !ok {
+		req.Labels[shared.BackendConditionIdLabel] = req.ConditionId.Id
+	}
+
+	if _, ok := req.Annotations[shared.BackendConditionNameLabel]; !ok {
+		req.Annotations[shared.BackendConditionNameLabel] = req.ConditionName
+	}
+
```
Comment (kralicky) [inline plugins/alerting/pkg/alerting/api.go:43]: Add a short comment describing the intention for this logic

File: plugins/alerting/pkg/alerting/api_conditions.go [hunk 4]
```diff
@@ -343,28 +261,35 @@ func (p *Plugin) AlertConditionStatus(ctx context.Context, ref *corev1.Reference
 		return nil, shared.WithInternalServerError("cannot parse response body into expected api struct")
 	}
 	for _, alert := range respAlertGroup {
+		matched := false
```
Comment (kralicky) [inline plugins/alerting/pkg/alerting/api_conditions.go:264]: This logic is somewhat hard to follow. Maybe refactor such that the matched variable isn't needed

File: plugins/alerting/pkg/alerting/api_conditions.go [hunk 5]
```diff
@@ -515,60 +477,59 @@ func (p *Plugin) Timeline(ctx context.Context, req *alertingv1.TimelineRequest)
 	cortexStep := durationpb.New(req.LookbackWindow.AsDuration() / 500)
 	var wg sync.WaitGroup
 	var addMu sync.Mutex
-	for idx := range conditions {
-		idx := idx // capture in closure
+	for _, cond := range conditions {
```
Comment (kralicky) [inline plugins/alerting/pkg/alerting/api_conditions.go:480]: The synchronization logic here will have significant lock contention. I suggest refactoring this so that you don't need to lock in the goroutine, instead setting resp.Items all at once, at the end.

**Review by kralicky — CHANGES_REQUESTED** [review]

File: pkg/alerting/storage/storage_opts/opts.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	return &SyncOptions{
+		Timeout: 2 * time.Minute,
+		Router:  routing.NewDefaultOpniRouting(),
+	}
+}
+
+type ClientSetOptions struct {
+	Logger     *zap.SugaredLogger
+	Timeout    time.Duration
+	TrackerTTl time.Duration
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply pkg/alerting/storage/storage_opts/opts.go:63]: Done

File: pkg/alerting/storage/stores.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/rancher/opni/pkg/alerting/drivers/routing"
+	"github.com/rancher/opni/pkg/alerting/interfaces"
+	"github.com/rancher/opni/pkg/alerting/storage/storage_opts"
+	alertingv1 "github.com/rancher/opni/pkg/apis/alerting/v1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type AlertingStoreBroker interface {
+	NewClientSet() AlertingClientSet
+	// Use handles specify a specific store in the clientset.
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply pkg/alerting/storage/stores.go:15]: Clarified what `Use` is meant to do

File: pkg/alerting/storage/stores.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	// It is also responsible for migrating existing data if there is an existing store
+	// of the same type already in use by the clientset
+	Use(store any)
+}
+
+type AlertingClientSet interface {
+	Purge(ctx context.Context) error
+	HashRing
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply pkg/alerting/storage/stores.go:23]: Done

File: pkg/alerting/storage/stores.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
+package storage
+
+import (
+	"context"
+
+	"github.com/rancher/opni/pkg/alerting/drivers/routing"
+	"github.com/rancher/opni/pkg/alerting/interfaces"
+	"github.com/rancher/opni/pkg/alerting/storage/storage_opts"
```
[the remark this answers is not in the document]
Comment (alexandreLamarre) [reply pkg/alerting/storage/stores.go:8]: Refactor to `opts`

File: pkg/apis/alerting/v1/validate.go (hunk 1 above)
Comment (alexandreLamarre) [reply pkg/apis/alerting/v1/validate.go:24]: Done

File: plugins/alerting/pkg/alerting/api.go (hunk 2 above)
Comment (alexandreLamarre) [reply plugins/alerting/pkg/alerting/api.go:19]: Yep

File: plugins/alerting/pkg/alerting/api_conditions.go (hunk 4 above)
Comment (alexandreLamarre) [reply plugins/alerting/pkg/alerting/api_conditions.go:264]: Turns out this logic is correct iff len(router.HasLabels(conditionId)) == 1, so I'll rewrite & tidy up this portion

**Review by kralicky — APPROVED** [review]

**Review by kralicky — APPROVED** [review]
